lib: picolib: put stdio variables into libc partition

dcpleung opened the pull request:
This adds the necessary modifier to the stdin/stdout/stderr variables in picolib, and putting into the z_libc_partition. This allows userspace applications to utilize these variables for console I/O.

Fixes #51343

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
lib: picolib: put stdio variables into libc partition

This adds the necessary modifier to the stdin/stdout/stderr
variables in picolib, and putting into the z_libc_partition.
This allows userspace applications to utilize these variables
for console I/O.

Fixes #51343

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Compliance check complains about `static` is not the first of the line. Though it is following the style of the file.

keith-packard reviewed (dismissed):
Looks like this isn't strictly required as linker scripts use a filename pattern that matches the hooks file as well as the library itself, but it's a good cleanup.

keith-packard commented:
> Compliance check complains about `static` is not the first of the line. Though it is following the style of the file.

Feel free to fix the file to make compliance happy

dcpleung committed:
lib: picolib: move static to be the first modifier

Compliance check complains about static not being the first
modifier. So move them so there are no more complains
in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
> Looks like this isn't strictly required as linker scripts use a filename pattern that matches the hooks file as well as the library itself, but it's a good cleanup.

Actually, those need to be in the z_libc_partition so userspace applications can access them. The entries in the linker script put them in the `.data` section which is inaccessible by userspace apps.

> > Compliance check complains about `static` is not the first of the line. Though it is following the style of the file.
> Feel free to fix the file to make compliance happy

Added another commit to move the modifier. Hopefully the check is happy now.

keith-packard commented:
> > Looks like this isn't strictly required as linker scripts use a filename pattern that matches the hooks file as well as the library itself, but it's a good cleanup.
> 
> Actually, those need to be in the z_libc_partition so userspace applications can access them. The entries in the linker script put them in the `.data` section which is inaccessible by userspace apps.

Yeah, all of my testing has shown that the generated linker scripts (save qemu_x86_tiny) end up with this:
```
  z_data_smem_z_libc_partition_part_start = .;
  KEEP(*(data_smem_z_libc_partition_data*))
  *libc.a:*(.data .data.* .sdata .sdata.*)
  z_data_smem_z_libc_partition_bss_start = .;
  KEEP(*(data_smem_z_libc_partition_bss*))
  *libc.a:*(.bss .bss.* .sbss .sbss.* COMMON COMMON.*)
  z_data_smem_z_libc_partition_bss_end = .;
  . = ALIGN(0x1000);;
  z_data_smem_z_libc_partition_part_end = .;
```
the `*libc.a:*` pattern ends up matching `liblib__libc__picolibc.a(libc-hooks.c.obj)` so the hooks end up in z_libc_partition by accident. This doesn't work for newlib because `libc-hooks.c` is in `liblib__libc__newlib.a`, which doesn't match `*libc.a:*`. Now that I read this, it's clear that the declarations in libc-hooks.c should be decorated as you've indicated and that they have only worked by mere coincidence. I'm kinda curious how you discovered this problem? Is there some test for which this actually fails?

> Added another commit to move the modifier. Hopefully the check is happy now.

Sounds good. I'm always happy to abide by local code format requirements, sorry I missed some here!

keith-packard reviewed (approved):
Looks great!

dcpleung commented:
> I'm kinda curious how you discovered this problem? Is there some test for which this actually fails?

This is mostly a manual process. Once I found out they are global variables (via addresses in exception message), the solution is to put them into memory partitions so that user threads can access them. As they are C library variables, they need to be in the z_libc_parition.
